### ISSUE TITLE
Widget representation in the view

### ISSUE THREAD

**Issue by szymonkups** [ckeditor/ckeditor5-engine#693]
Currently we have three specialized types of elements in the view:
- `AttributeElement`,
- `ContainerElement`,
- `EmptyElement`.

Now we want to introduce widgets. Things we know so far that are needed to properly represent them in the view:

- we need to be able to tell if given view element is a widget,
- we need to be able to tell what kind of widget it is (image, table, etc.),
- we need to be able to represent widgets differently for data and editing pipelines.

### WidgetElement class

First idea was to introduce `WidgetElement` class extending `ContainerElement`.
It would be used as a base class for all future widgets. It could add proper `getFillerOffset`,  CSS classes and set `contenteditable` attribute to `false`. Finding if element is a widget and what kind of widget it is would be very easy, by using `instanceof` operator.

Problem with this approach is that `WidgetElement` class would only be valid for editing pipeline. For data pipeline each widget is just a `ContainerElement`.

### widgetize() decorator

In this approach we could define widgets using `ContainerElements`, and represent them as they should appear in data pipeline. For editing pipeline, widgets would be decorated with `widgetize` function:
```
createImageWidget(); // <figure class="image"><img src="..." alt="..." /></figure>
widgetize( createImageWidget() ); // <figure class="image ck-widget" contenteditable="false"><img src="..." alt="..." /></figure>
```
This way we could avoid creation of new sub-classes, and show that representation in editing pipeline just "decorates" data representation.

But how to distinguish widget elements from each other? `widgetize()` could add some custom properties to decorated element:
```
function widgetize( viewElement ) {
    viewElement.setAttribute( 'contenteditable', false );
    viewElement.addClass( 'ck-widget' );
    viewElement.isWidget = true;
}
```
but those custom properties will not be present in cloned elements, so this approach is not totally valid. 

Maybe we could add some meta-information to `ContainerElements` about widget type as some special attributes, that would be cloned as regular attributes but will not be rendered?

**Comment by Reinmar** [comment]
> Problem with this approach is that WidgetElement class would only be valid for editing pipeline. For data pipeline each widget is just a ContainerElement.

To clarify why it's a problem. If anyone would like to create a specialised subclass like `ImageElement`, they would need to make two subclasses – `ImageElement` for the data pipeline and `ImageWidgetElement` for the editing pipeline.

**Comment by Reinmar** [comment]
The problem with `widgetize()` is that the `isWidget` property won't be cloned by `clone()`. Also, an overridden `getFillerOffset()` method would be lost.

Therefore, we need two changes:

1. Ability to specify `isWidget` in a way that it won't be lost on cloning. We've got two ideas:
  * add special meta attributes to view elements (verbose),
  * have special meta attributes support built into the `get/setAttribute()` methods of the `Element` class. One idea is that `setAttribute( Symbol( 'isWidget' ), true )` could have special behaviour – such attributes wouldn't be written to DOM. We can also use special `data-*` attributes which would be written to DOM – it's not a big issue in editing pipeline but would be inconvenient for the data pipeline when you don't want to have unnecessary attributes.
2. Make `getFillerOffset()` a real behaviour, not a class method. The `Element`'s subclasses constructors could set those properties and the `widgetize()` method would do that too. Then, the `clone()` method would need to copy it. (Reminder – please have just one reference to each kind of this behaviour function.)

BTW. We're talking about classes like `ImageElement` (we already have `LinkAttributeElement`). We could avoid having even these subclasses by using the meta attributes to mark the special "type" of the element.

**Comment by pjasiun** [comment]
> > Problem with this approach is that WidgetElement class would only be valid for editing pipeline. For data pipeline each widget is just a ContainerElement.
>
> To clarify why it's a problem. If anyone would like to create a specialised subclass like ImageElement, they would need to make two subclasses – ImageElement (extends ContainerElement) for the data pipeline and ImageWidgetElement (extends WidgetElement) for the editing pipeline.

I do not get why someone may want to create any specific subclass for the data pipeline.

**Comment by pjasiun** [comment]
> Therefore, we need two changes:
> ...

I'm not sure if you do not overengineer it. Can't simple `clone()` check if `isWidget` is true and clone it if it is true?

**Comment by Reinmar** [comment]
> I do not get why someone may want to create any specific subclass for the data pipeline.

I guess no one needs that... until someone will (e.g. to have a special filler behaviour). But that's not the point in fact.

The point is that we must strive to make the converters code as clean as possible. If a converter for the editing pipeline must return an `ImageElement` and the converter for data `ContainerElement`, then it's an unnecessary complication.

**Comment by Reinmar** [comment]
> I'm not sure if you do not overengineer it. Can't simple clone() check if isWidget is true and clone it if it is true?

It'd be a god if I were sure that we're not overengineering it, but my point is that we're de-hardcoding the thing that may change.

There are piece of code which are unlikely to change and there are pieces which may eventually change. The goal is always to identify the latter as soon as possible and hide them from the things which are constant.

The `Element` class itself is that constant thing – it will be there forever. However, concrete implementations like widgets, image, link feature, etc. are a subject to change, and, most importantly, we'll never guess all possible use cases.

So, we can hardcode now that the `Element` class knows about `isWidget`. Easy. But the `Element` class should not know such things like widgets cause they are a specific use case and may change. For example – what if we'll also need `isInlineWidget`, `widgetType`, `isSuperSpecialSomething`? You can extend base `clone()` with all that (especially if you consider 3rd party code).

So what do is to extract the thing that may change (so special meta attributes) and hide them beneath an interface (in form of a special meta properties).

**Comment by pjasiun** [comment]
I would go with `ContainerElement` knowing about the widget. Even if `getFillerOffset` check `isWidget` it is fine for me. Later, when we will have the working code we may consider what and how we want to extract.

Also, I do not expect many features like widgets soon. Of course, we should create a flexible API, but  we do not need to thread widgets like the of the many features of that type. At least for now.

**Comment by pjasiun** [comment]
So for me, the first implementation can be with the logic inside `ContainerElement` class.

Also, I expect that there will be a special conversion builder: `fromElemenet().toWidget()`, or some kind of helper, but as I see the title it is not a scope of this issue.

**Comment by Reinmar** [comment]
> Later, when we will have the working code we may consider what and how we want to extract.

We have a working code now. We did this refactoring based on the image feature implementation.

> but we do not need to treat widgets like the of the many features of that type. At least for now.

We have 10% of the first widget so far, nothing even closely resembling tables with rich drag handlers, "add/remove row/col", "select row/col" widgets, no inline widgets, no placeholder, no magicline and no requests from the community. What's more – we won't have many of these things before 1.0. So how long is that "for now"?

Most importantly, though, what's wrong with custom properties? We had them in CKE4, we used `data-*` attributes in CKE4 and use in CKE5 inside the editable. We'd use `data-widget` like we used in CKE4, but that's a problem for the data pipeline, because this attr should not be rendered there. So we understood that there's a need for special, non-renderable attributes.

We could go with `isWidget` and some hacks, but there's nothing wrong in noticing that what we actually want are custom, non-renderable attributes. It's a harmless generalisation based on experience from CKE4.
